### DISCUSSION TITLE
Seeing extra leading character '[' in command output

### DISCUSSION THREAD

**Discussion by jharshman** [ghostty-org/ghostty#3566] [Q&A] [answered]
After compiling from source and running I am seeing an extra character in the output of every command. Is anyone else experiencing this? Shell is Bash and no modifications to the PS1. 

```bash
[josh@serenity ghostty]$ echo hi
[hi
[josh@serenity ghostty]$ pwd
[/home/josh/Projects/ghostty
```

```
info: ghostty version=1.0.1-dev+0000000
info: ghostty build optimize=ReleaseFast
info: runtime=apprt.Runtime.gtk
info: font_backend=font.main.Backend.fontconfig_freetype
info: dependency harfbuzz=8.4.0
info: dependency fontconfig=21402
info: renderer=renderer.OpenGL
info: libxev backend=main.Backend.io_uring
info(os): setlocale from env result=en_US.UTF-8
info(gtk): GTK version build=4.12.4 runtime=4.16.5
```

---
**UPDATE:**
An issue has been created for this bug: https://github.com/ghostty-org/ghostty/issues/4494

**Comment by jmckillican** [comment]
is this on your local machine or when ssh'd?

**Reply by jharshman** [reply]
On local

**Reply by rhodes-b** [reply]
@jparise Ive also seen this with NixOS bash shell integation after the fix with the prompt

**Reply by jparise** [reply]
@rhodes-b assume you mean #3003, yes? I suspected that was related, too, but I don't see how it would be producing a stray `[` given that the only use of that character there is escaped (`\[`).

It also doesn't look like your use case should fall into the multiline prompt category (e.g. no `\n` in your original prompt).

But let's keep looking at it.

**Comment by jparise** [comment]
This has been reported a few times, and we've been having trouble reproducing it. I definitely want to get it fixed though!

A few questions to help me narrow it down:

1. Which bash version are you using?
2. Does this still happen when shell integration is disabled (`shell-integration = none`)?
3. Which shell integration features do you have enabled/disabled (`shell-integration-features`)?
4. Could you share the value of your `$PS0`, `$PS1`, and `$PS2` prompt environment variables? (`$PS4` is also relevant if tracing somehow got enabled, but that seems unlikely.)

**Reply by rhodes-b** [reply]
1. For my setup this does not happen with `shell-integration = none`
2. I did not test disabling / enabling any features specific to shell integation

`$PS0` = ` ` <-- empty
`$PS1` = `\n\[\033[1;32m\][\[\e]0;\u@\h: \w\a\]\u@\h:\w]\$\[\033[0m\]`
`$PS2` = `>`
`$PS3` = ` ` <-- empty
`$PS4` = `+`

**Reply by jparise** [reply]
@rhodes-b There _is_ a newline in that prompt. I wonder if the stray `[` goes away if you remove it by adjusting your PS1 to:

```bash
PS1='\[\033[1;32m\][\[\e]0;\u@\h: \w\a\]\u@\h:\w]\$\[\033[0m\] '
```

The newline does explain how we'd be entering the code path from #3003, but I still don't yet see how that could result in a `[` being printed (and I can't reproduce what you're seeing yet using your origin prompt with the newline).

**Reply by rhodes-b** [reply]
> The newline does explain how we'd be entering the code path from #3003, but I still don't yet see how that could result in a `[` being printed (and I can't reproduce what you're seeing yet using your origin prompt with the newline).

setting to what you suggested above does seem to work but ofc has a side effect of no more empty line when refreshing the prompt

**Reply by jparise** [reply]
And how about this version, which simplifies it a bit by removing the embedded window title sequence? The empty line is restored and the color remains.

```bash
PS1='\n\[\033[1;32m\][\u@\h:\w]\$\[\033[0m\] '
```

**Reply by jparise** [reply]
I _think_ the problem might be with the original (nixos-defined?) prompt in this case, which nests the window title sequence within a printable pair of brackets.

```bash
PS1='\n\[\033[1;32m\][\[\e]0;\u@\h: \w\a\]\u@\h:\w]\$\[\033[0m\] '
                     ^-- open             close --^
```

I wonder whether writing it like this might be less likely to trigger the issue you're seeing:

```bash
PS1='\n\[\033[1;32m\]\[\e]0;\u@\h: \w\a\][\u@\h:\w]\$\[\033[0m\] '
```

Or even moving that window title sequence before all of the visible bits:

```bash
PS1='\[\e]0;\u@\h: \w\a\]\n\[\033[1;32m\][\u@\h:\w]\$\[\033[0m\] '
```

Note that we do our own window title capturing (unless the `title` shell integration feature is disabled), so you might not want/need that sequence in your prompt at all. That's totally up to you, and we should support it either way. Just an FYI.

**Reply by rhodes-b** [reply]
> I _think_ the problem might be with the original (nixos-defined?) prompt in this case, which nests the window title sequence within a printable pair of brackets.

yeah it comes from `/etc/bashrc`
```sh
# Provide a nice prompt if the terminal supports it.
if [ "$TERM" != "dumb" ] || [ -n "$INSIDE_EMACS" ]; then
  PROMPT_COLOR="1;31m"
  ((UID)) && PROMPT_COLOR="1;32m"
  if [ -n "$INSIDE_EMACS" ]; then
    # Emacs term mode doesn't support xterm title escape sequence (\e]0;)
    PS1="\n\[\033[$PROMPT_COLOR\][\u@\h:\w]\\$\[\033[0m\] "
  else
    PS1="\n\[\033[$PROMPT_COLOR\][\[\e]0;\u@\h: \w\a\]\u@\h:\w]\\$\[\033[0m\] "
  fi
  if test "$TERM" = "xterm"; then
    PS1="\[\033]2;\h:\u:\w\007\]$PS1"
  fi
fi
```

`PS1='\n\[\033[1;32m\][\u@\h:\w]\$\[\033[0m\] '` does not work
`PS1='\n\[\033[1;32m\]\[\e]0;\u@\h: \w\a\][\u@\h:\w]\$\[\033[0m\] '` does not work
`PS1='\[\e]0;\u@\h: \w\a\]\n\[\033[1;32m\][\u@\h:\w]\$\[\033[0m\] '` does not work

I tried this by `unset PS1` entering the prompt a few times with then export those options above `[` still persists

**Reply by rhodes-b** [reply]
Another weird thing I just noticed is that it only happens if I invoke ghostty from cli. If I spawn the exe from file explorer it doesnt happen is shell integration not happening in that case?

**Reply by jparise** [reply]
Here's some information on how to very shell integration: https://ghostty.org/docs/features/shell-integration#verify-shell-integration

The other other suggestion I can offer is to try experimenting with some local changes to `ghostty.bash` (`$GHOSTTY_RESOURCES_DIR/shell-integration/bash/ghostty.bash`) to narrow things down further.

I'm afraid I'm otherwise at a loss at understanding what's going on in your environment that would cause this. Hopefully some other detail will pop up that points to the root cause.

**Reply by jharshman** [reply]
For what it's worth I am also launching it from the CLI when I see this behavior.

```bash
josh@serenity:~$ echo $PS0
\e]133;C\e\\\r
josh@serenity:~$ echo $PS1
\[\e]133;D;$?\e\\\e]133;A\e\\\]${PROMPT_START@P}\[\e[${PROMPT_COLOR}${PROMPT_HIGHLIGHT:+;$PROMPT_HIGHLIGHT}m\]${PROMPT_USERHOST@P}\[\e[0m\]${PROMPT_SEPARATOR@P}\[\e[${PROMPT_DIR_COLOR-${PROMPT_COLOR}}${PROMPT_HIGHLIGHT:+;$PROMPT_HIGHLIGHT}m\]${PROMPT_DIRECTORY@P}\[\e[0m\]${PROMPT_END@P}\$\[\e[0m\] \[\e]133;B\e\\\]
josh@serenity:~$ echo $PS2 
>
josh@serenity:~$ echo $PS3

josh@serenity:~$ echo $PS4
+

```

**Reply by vhodges** [reply]
@jparise I am also seeing this. Installed via the Nix flake.  Note,  disabling shell integration entirely 'fixes' the issue.  fwiw, I am running Bash version 5.2.37.  

I thought too it was prompt related but I tried setting PS1 and the [ still was showing up (but it's been a long time since I tried changing my prompt and so might have done something wrong).  Maybe related to the pre-run functions?  I tried following what the integration code was doing but my bash fu was not strong enough.

**Reply by rhodes-b** [reply]
I havent looked into this like at all, I should do that Im ok with bash at the same time im also ok with `shell-integration = none` since the features like cwd on new tabs and updating the title already work for me without it

**Reply by jparise** [reply]
I _really_ wish I could reproduce this. I unfortunately don't have time to set up a NixOS installation myself. (It doesn't seem that easy to spin an equivalent environment up within a Docker container, right?)

I'm getting close to just doing something like this (half joking / half not) in the meantime because I can't think of another way to detect the specific root cause:

```diff
diff --git i/src/shell-integration/bash/ghostty.bash w/src/shell-integration/bash/ghostty.bash
index 72ae455d..bc11954c 100644
--- i/src/shell-integration/bash/ghostty.bash
+++ w/src/shell-integration/bash/ghostty.bash
@@ -128,7 +128,7 @@ function __ghostty_precmd() {
       # mark the last line as a secondary prompt (k=s) to prevent the
       # preceding lines from being erased by ghostty after a resize.
       if [[ "${PS1}" == *"\n"* || "${PS1}" == *$'\n'* ]]; then
-        PS1=$PS1'\[\e]133;A;k=s\a\]'
+        [ ! -f /etc/NIXOS ] && PS1=$PS1'\[\e]133;A;k=s\a\]'
       fi
 
       # Cursor
```

**Reply by jharshman** [reply]
> It doesn't seem that easy to spin an equivalent environment up within a Docker container, right?

It's easy to do so with a Docker container. Just have to install the deps and zig == 0.13 and then run zig build.

**Reply by jparise** [reply]
> > It doesn't seem that easy to spin an equivalent environment up within a Docker container, right?
> 
> It's easy to do so with a Docker container. Just have to install the deps and zig == 0.13 and then run zig build.

I meant a container environment that reproduces the problem, not a container for building Ghostty.

**Reply by jharshman** [reply]
> I meant a container environment that reproduces the problem, not a container for building Ghostty.

Oh then no you will want a VM.

**Comment by ghost** [comment]
I can easily reproduce this on fedora 41. If I launch it from an existing shell it looks like this [https://pasteboard.co/dR7JtAjeq3vF.png](https://pasteboard.co/dR7JtAjeq3vF.png) and if I launch it by double clicking on it from the file explorer it looks like this [https://pasteboard.co/Laqq8seVxMH8.png](https://pasteboard.co/Laqq8seVxMH8.png)

Note that I get a slightly different problem here, which is because I use shell-integration-features = no-cursor. If I don't use that line then I see the [ that's reported in this thread. But as described above, something appears to be going wrong with shell injection if run from an existing shell.

**Reply by jparise** [reply]
> I can easily reproduce this on fedora 41. If I launch it from an existing shell it looks like this [pasteboard.co/dR7JtAjeq3vF.png](https://pasteboard.co/dR7JtAjeq3vF.png) and if I launch it by double clicking on it from the file explorer it looks like this [pasteboard.co/Laqq8seVxMH8.png](https://pasteboard.co/Laqq8seVxMH8.png)

I believe this is due to shell integration only (automatically) applying to the second case, which is consistent with how others are experiencing this problem.

**Reply by jparise** [reply]
@mattxtaz does your prompt (`PS1`) contain a newline? It doesn't look like it from your screenshots, which would eliminate the "multiline prompt" path that we noted above.

**Reply by ghost** [reply]
No it does not. It's the basic "out of the box" prompt that comes with fedora. I've not changed it at all. It is using the "color-bash-prompt" that comes with fedora though which does a lot of stuff to make the prompt look like the one that comes with ubuntu. Although in those screenshots it's using the traditional prompt because the $TERM ends in ghostty and not in color. But anybody ought to be able to reproduce it on fedora quite easily. Just open gnome terminal and then run the ghostty binary from that shell.

Also, I've tried creating a simple shell script which simply has bash as its interpreter and then runs ghostty from that. Double clicking on that from the gnome file explorer shows exactly the same problem. And double clicking on the actual ghostty binary there is no problem at all. So it's launching it from another shell that causes the problem.

**Reply by jparise** [reply]
I tried out [`color-bash-prompt`](https://copr-dist-git.fedorainfracloud.org/cgit/petersen/bash-color-prompt/bash-color-prompt.git/tree/bash-color-prompt.sh) locally on macOS, but that unfortunately wasn't enough to trigger the issue for me.

**Reply by jharshman** [reply]
Since we have identified the issue and have multiple people able to replicate it, let's create an issue for this problem if it doesn't already exist.

**Reply by jparise** [reply]
Just to continue recording my notes, I also haven't been able to reproduce this using a Fedora container (under macOS) with a basic system setup with and without color-bash-prompt. No windowing environment (e.g. gnome) though.

At this point, the most helpful thing would be narrowing down _where_ the stray `[` is originating. It seems like it's coming from somewhere (in)directly from `ghostty.bash`, but I'm just not seeing where yet. I've probably just been looking at it too long because it seems like it should be obvious.

I also agree it makes sense to create an issue for this. I'll do that now.

**Comment by anund** [comment]
Since the issue seems to be machine configuration dependent it would be helpful to generate a machine with the issue. For those in using nix can you please help identify a configuration that reliably triggers the issue? I cannot reproduce the issue just swapping my shell for bash in nix. 

To do that here's a [gist](https://gist.github.com/anund/cd75dcf90e15776805507a550d96227e) running ghostty in a vm with instructions attached to the gist. The `system-config` attribute should be changed to match your system enough to reproduce the issue. The vm _does not_ currently reproduce the issue. (the plasma config is not required, it just an example)

**Reply by anund** [reply]
@vhodges Stab at a cinnamon de env. No change on reproducing :nuts:
```nix
    # services.displayManager.sddm.enable = true;
    # services.displayManager.sddm.wayland.enable = true;
    # services.desktopManager.plasma6.enable = true;

    services.xserver = {
        enable = true;
        displayManager.lightdm.enable = true;
        desktopManager.cinnamon.enable = true;
    };
    services.libinput.enable = true;
    services.displayManager.defaultSession = "cinnamon";
```

**Reply by vhodges** [reply]
I don't think vte.sh is being sourced -  have no idea why :-/ 

but If you ```source /nix/store/*vte-*/etc/profile.d/vte.sh``` before sourcing  test-vte.sh, you'll see the issue.

**Reply by anund** [reply]
Thank you. I looked again and had found a partial misunderstanding on my part. The extra `[` is only printed when the previous command has output. Mashing enter doesn't show anything. That makes sense given you narrowed it down to cursor interactions. I can reproduce the issue now. (either with the script in `gnome-terminal` or via launching `ghostty` _from_ `gnome-terminal`)


@jharshman you likely have the same issue as vhodges (see below). @rhodes-b you might or might not have the issue depending on how you're launching ghostty.

```
printf '%s' "$PS0" -> \e]133;C\e\\\r
printf '%q' "$PS0" -> $'\\e]133;C\\e\\\\\\r\E]666;vte.shell.preexec!\E\\'
```

@jparise trick for the future.

**Reply by jparise** [reply]
Relatedly, I'm also attempting to reduce our reliance on `PS0`: https://github.com/ghostty-org/ghostty/pull/4775

**Comment by jharshman** [comment]
Cross linking issue for completeness.
https://github.com/ghostty-org/ghostty/issues/4494

**Comment by pbulteel** [comment]
This seems like the more aligned bug to my issue, but I see more than just [

https://i.imgur.com/UZbebmj.png

Ghostty 1.0.1

Version
  - version: 1.0.1
  - channel: stable
Build Config
  - Zig version: 0.13.0
  - build mode : builtin.OptimizeMode.ReleaseFast
  - app runtime: apprt.Runtime.gtk
  - font engine: font.main.Backend.fontconfig_freetype
  - renderer   : renderer.OpenGL
  - libxev     : main.Backend.io_uring
  - GTK version:
    build      : 4.6.9
    runtime    : 4.6.9
  - libadwaita : enabled
    build      : 1.1.7
    runtime    : 1.1.7
  - libX11     : enabled

I'm using the ghossty from github mkasberg/ghostty-ubuntu  to run it on Pop!_OS (Based on Ubuntu 22.04) - I have another system with Ubuntu 24.04 LTS that I have only just installed and therefore haven't used. I'll check what is happening there. What other details would you need from me to help narrow down this bug?

**Comment by vhodges** [comment, marked as answer]
I can confirm 1.0.2 fixes the issue I was seeing.
